source/slang/slang-emit-cpp.cpp

```diff
@@ -607,29 +641,42 @@ IRInst* CPPSourceEmitter::_clone(IRInst* inst)
         return nullptr;
     }
 
+    IRModule* module = inst->getModule();
+    // All inst's must belong to a module
+    SLANG_ASSERT(module);
+
     // If it's in this module then we don't need to clone
-    if (inst->getModule() == m_uniqueModule)
+    if (module == m_uniqueModule)
     {
         return inst;
     }
 
-    if (isNominalOp(inst->op))
-    {
-        // If it's nominal we don't bother copying, as we assumed it is already de-duped
-        return inst;
-    }
-    
     if (IRInst*const* newInstPtr = m_cloneMap.TryGetValue(inst))
     {
         return *newInstPtr;
     }
-    
+
+    if (isNominalOp(inst->op))
+    {
+        // TODO(JS)
+        // This is arguably problematic - I'm adding an instruction from another module to the map, to be it's self.
+        // I did have code which created a copy of the nominal instruction and name hint, but because nominality means
+        // 'same address' other code would generate a different name for that instruction (say as compared to being a member in
+        // the original instruction)
+        //
+        // Because I use findOrAddInst which doesn't hoist instructions, the hoisting doesn't rely on parenting, that would
+        // break.
```

Comment:
The right solution is to migrate away from this entire approach to handling CPU codegen (which involves synthesizing a whole lot of code/types during emit), which is what we've discussed before. I'm okay with whatever hacks we need to make to get this done for now, but I really don't want us to keep digging ourselves in deeper in terms of effort spent on this approach as opposed to more general solutions that could benefit this and other back-ends (namely, putting as many of the definitions as possible directly into the Slang stdlib code and letting the existing codegen machinery do its job).